ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/SlotStore.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2020, Red Hat, Inc. and/or its affiliates,
+ * and individual contributors as indicated by the @author tags.
+ * See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ * This copyrighted material is made available to anyone wishing to use,
+ * modify, copy, or redistribute it subject to the terms and conditions
+ * of the GNU Lesser General Public License, v. 2.1.
+ * This program is distributed in the hope that it will be useful, but WITHOUT A
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License,
+ * v.2.1 along with this distribution; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA  02110-1301, USA.
+ *
+ * (C) 2020,
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com)
+ */
+package com.arjuna.ats.internal.arjuna.objectstore.slot;
+
+import com.arjuna.ats.arjuna.logging.tsLogger;
+import com.arjuna.ats.arjuna.objectstore.StateStatus;
+import com.arjuna.ats.arjuna.state.InputBuffer;
+import com.arjuna.ats.arjuna.state.InputObjectState;
+import com.arjuna.ats.arjuna.state.OutputObjectState;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+/**
+ * A storage system presenting a key-value API, implemented (conceptually) using a fixed sized array.
+ *
+ * @author Jonathan Halliday (jonathan.halliday@redhat.com), 2020-03
+ */
+public class SlotStore {
+
+    private final SlotStoreEnvironmentBean config;
+
+    private final String storeDirCanonicalPath;
+
+    /*
+     * The slotIdIndex tracks the key to slot mapping for all in-use slots,
+     * whilst the free list tracks all unoccupied slots. Care must be taken with the
+     * relative ordering of operations on these structures, to avoid losing or double allocating slots.
+     */
+    public final ConcurrentHashMap<SlotStoreKey, Integer> slotIdIndex = new ConcurrentHashMap<>();
+    public final Deque<Integer> freeList = new ConcurrentLinkedDeque<>();
+
+    public final RAMSlots slots; // TODO add alternative implementations
+
+    /**
+     * Create a new instance with the given configuration.
+     *
+     * @param config The configuration parameters for the instance
+     * @throws IOException if the required backing storage can't be initialized.
+     */
+    public SlotStore(SlotStoreEnvironmentBean config) throws IOException {
+
+        this.config = config;
+
+        // unused for now, but eventually we'll have a disk backend...
+        File storeDir = new File(config.getStoreDir());
+        if (!storeDir.exists() && !storeDir.mkdirs()) {
```

Comment:
Can the file exist if the directory hierarchy hasn't been created?

Reply:
The dir may be created in one run and its existence validated on the next, since it doesn't get deleted when the server shuts down. Wouldn't be a lot of use for recovery if it did! That code shouldn't really be in the RAM backed version at all, but the path also serves as the store name so it didn't get cleanly factored out.

first run of a server, exist check fails, so mkdirs runs and must succeed
subsequent runs of a server, exists check passes, so mkdirs doesn't run as it's shortcut out

Reply:
Of course, thanks for the clarification.